### PR TITLE
Fix memory leak while compression

### PR DESCRIPTION
**Pull request by mhmtszr**
## Links
We have been observing the memory increase in our Go applications for a long time. When We look at the cumulative increase of the memories used by the application with the memory dump method, We observed that the New Relic agent uses a lot of memory while doing gzip compression.

As a result, too much GC is running. You can visualize it by closing GOGC=off and defining GOMEMLIMIT this will avoid unnecessary GC, and will trigger the GC only if it's closed to the GOMEMLIMIT.

Issue: https://github.com/newrelic/go-agent/issues/619

## Details
Instead of creating a new gzip writer for every collect request, use common ones by using gzip writer sync pool.

### CODE REVIEW

File: v3/newrelic/collector.go [hunk 1]
```diff
@@ -136,11 +137,10 @@ func rpmURL(cmd rpmCmd, cs rpmControls) string {
 	return u.String()
 }
 
-func compress(b []byte) (*bytes.Buffer, error) {
+func compress(b []byte, w *gzip.Writer) (*bytes.Buffer, error) {
 	var buf bytes.Buffer
-	w := gzip.NewWriter(&buf)
+	w.Reset(&buf)
 	_, err := w.Write(b)
-	w.Close()
```
Comment (mstrYoda) [inline v3/newrelic/collector.go:143]: You need to close gzip writer to read that data properly. It adds required gzip headers.

File: v3/newrelic/collector.go [hunk 2]
```diff
@@ -136,11 +137,10 @@ func rpmURL(cmd rpmCmd, cs rpmControls) string {
 	return u.String()
 }
 
-func compress(b []byte) (*bytes.Buffer, error) {
+func compress(b []byte, w *gzip.Writer) (*bytes.Buffer, error) {
```
Comment (mstrYoda) [inline v3/newrelic/collector.go:140]: If you access to a pointer gzip writer concurrently, it will probably cause race condition.

**Review by mstrYoda — DISMISSED** [review]

**Comment by mhmtszr** [comment]
@iamemilio this memory leak affects our applications, could you please review it?

Also, I realized that **newAnalyticsEvents** function creates a lot of analytic events on the heap too. We may focus there too.

**Comment by iamemilio** [comment]
Would you be able to re-factor this for 1.17, that is the minimum supported version of Go that we must support currently so we can't use `any` yet unfortunately.

**Comment by mhmtszr** [comment]
@iamemilio I changed any to interface{} and I tested it works with 1.17 could you check it?

**Comment by iamemilio** [comment]
It looks like some of the tests are failing: https://github.com/newrelic/go-agent/actions/runs/3886936188/jobs/6669443533

Our CI is a bit chaotic, I have a fix coming through that will clean it up. I would focus on getting this test to pass for now.

**Comment by mhmtszr** [comment]
I thought I solved the failed tests could you run the pipelines again I can't.

**Comment by iamemilio** [comment]
Thanks for fixing this up, and sorry about the confusing CI. This looks good now! We are going to run some scale tests on our end just to verify, but it looks like this will very likely make it into the next release! Thanks for contributing.

**Review by iamemilio — APPROVED** [review]
Confirming a noticeable decrease in CPU and Memory usage

**Pull request merged**